common/src/main/java/com/bakdata/quick/common/api/client/PartitionedMirrorClient.java

```diff
@@ -0,0 +1,182 @@
+     * @param mirrorHost      host to use
+     * @param client          http client
+     * @param keySerde        the serde for the key
+     * @param valueResolver   the value's {@link TypeResolver}
+     * @param partitionFinder strategy for finding partitions
+     */
+    public PartitionedMirrorClient(final String topicName, final MirrorHost mirrorHost, final HttpClient client,
+                                   final Serde<K> keySerde, final TypeResolver<V> valueResolver,
+                                   final PartitionFinder partitionFinder) {
+        this.streamsStateHost = StreamsStateHost.fromMirrorHost(mirrorHost);
+        this.client = client;
+        this.keySerde = keySerde;
+        this.topicName = topicName;
+        this.partitionFinder = partitionFinder;
+        this.parser = new MirrorValueParser<>(valueResolver, client.objectMapper());
+        this.requestManager = new DefaultMirrorRequestManager(client);
+        initRouter();
+    }
+
+    @Override
+    @Nullable
+    public V fetchValue(final K key) {
+        final MirrorHost currentKeyHost = router.getHost(key);
+        return this.requestManager.sendRequest(Objects.requireNonNull(currentKeyHost).forKey(key.toString()),
+            this.parser::deserialize);
+    }
+
+    @Override
+    public List<V> fetchAll() {
+        final List<V> valuesFromAllHosts = new ArrayList<>();
+        for (final MirrorHost host : this.knownHosts) {
+            final List<V> valuesFromSingleHost =
+                Objects.requireNonNullElse(this.requestManager.sendRequest(host.forAll(), this.parser::deserializeList),
+                    Collections.emptyList());
+            valuesFromAllHosts.addAll(valuesFromSingleHost);
+        }
+        return valuesFromAllHosts;
+    }
+
+    @Override
+    @Nullable
+    public List<V> fetchValues(final List<K> keys) {
+        return keys.stream().map(this::fetchValue).collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean exists(final K key) {
+        return this.fetchValue(key) != null;
+    }
+
+
+    private void initRouter() {
+        log.info("Initializing partition router for the mirror at: {} and the topic: {}.",
+            this.streamsStateHost.getHost(), this.topicName);
+        final Map<Integer, String> response = makeRequestForPartitionHostMapping();
+        this.router = new PartitionRouter<>(this.keySerde, this.topicName, this.partitionFinder, response);
+        this.knownHosts = this.router.getAllHosts();
+    }
+
+    private Map<Integer, String> makeRequestForPartitionHostMapping() {
+        final String url = this.streamsStateHost.getPartitionToHostUrl();
+        try (final ResponseBody responseBody = requestManager.makeRequest(url)) {
+            if (responseBody != null) {
+                final Map<Integer, String> partitionHostMappingResponse = this.client.objectMapper()
+                    .readValue(Objects.requireNonNull(responseBody).byteStream(), MAP_TYPE_REFERENCE);
```

Comment:
```suggestion
                    .readValue(responseBody.byteStream(), MAP_TYPE_REFERENCE);
```
You've checked it above